include/onnxruntime/core/session/onnxruntime_cxx_api.h

```diff
@@ -88,6 +88,7 @@ template <typename T>
 #ifdef ORT_API_MANUAL_INIT
 const OrtApi* Global<T>::api_{};
 inline void InitApi() { Global<void>::api_ = OrtGetApiBase()->GetApi(ORT_API_VERSION); }
+inline void InitApi(const OrtApi* api) { Global<void>::api_ = api; }
```

Comment:
Need to discuss this.

Reply:
Can you comment on the motivation for this?

Reply:
To summarize, custom op dlls (like those in onnxruntime-extensions) are not linked to onnxruntime.dll and thus do not have access to ORT's `OrtGetApiBase()`. This means that calling `InitApi()` fails to link in custom ops. This is a problem because the C++ APIs rely on the global `Global<>::api_` object being set.

Custom ops are passed a pointer to an `OrtApiBase` object via their `RegisterCustomOps` export, which they can use to get a pointer to the `OrtApi` object and its contained functions.

So, in order to be able to use the newly updated C++ APIs, we need some way of setting this global `api_` object. Here's an example of how to call this new `InitApi(const OrtApi*)` overload from within a custom op dll.

Without this new overload of `InitApi`, custom dlls would only be able to use the C APIs or the now recently deprecated `CustomOpApi` struct.

```c++
// my_custom_op.cc

#define ORT_API_MANUAL_INIT
#include <onnxruntime_cxx_api.h>
#undef ORT_API_MANUAL_INIT

OrtStatus* ORT_API_CALL RegisterCustomOps(OrtSessionOptions* options, const OrtApiBase* api_base) {
  Ort::InitApi(api_base->GetApi(ORT_API_VERSION)); // This overload is new and does not exist yet.
  // ...
}
```

Here's a link to the discussion I had with Dmitri regarding the need for this new overload of `InitApi`: https://github.com/microsoft/onnxruntime/pull/12998#discussion_r975884735

We are still not sure this is the correct approach, but this is the simplest solution that works and allows me to create a prototype that uses new, non-deprecated C++ APIs.

Reply:
I think we need to put it in the header.

Reply:
I added a comment that explains when/how this is used.

Reply:
Would this be cleaner if it took a version number instead of an OrtApi*? That way it's far more equivalent with the existing InitApi()

Reply:
If you're suggesting `Ort::InitApi(api_base, ORT_API_VERSION)`, I agree that it is probably cleaner.

However, if you mean `Ort::InitApi(ORT_API_VERSION)`, that would not work because this is called custom op dlls that are not linked to onnxruntime.dll and thus do not have access to ORT's OrtGetApiBase().

Reply:
Re: OrtGetApiBase() being not available. You could just define it in your custom op library to do what you want. It doesn't need to be the one from onnxruntime.dll.

Then you don't need a third way to initialize this variable.

Reply:
How would this work? Perhaps like this?

``` C++
// my_custom_op.cc

#define ORT_API_MANUAL_INIT
#include <onnxruntime_cxx_api.h>
#undef ORT_API_MANUAL_INIT

const OrtApiBase* OrtGetApiBase() {
  return global_api_base;
}

static const OrtApiBase* global_api_base;

OrtStatus* ORT_API_CALL RegisterCustomOps(OrtSessionOptions* options, const OrtApiBase* api_base) {
  global_api_base = api_base;  // Required before using C++ APIs.
  // ...
}
```

I'm not sure this is easier to work with.

Reply:
By `you`, you mean the customer. I think we should not make customers life more complicated than already is.